Improve echo, read, mapfile documentation

dcghub opened the pull request:
I just wanted to check your opinion on these changes. I've tried to make texts as short and concise as possible, and I have tried to mark as unimplemented those that I _think_ they are.

The spec tests have been helpful, as you told me.

andychu commented on doc/help.md:
```diff
@@ -580,14 +580,70 @@ PS1=$x.
+
+<h4 id="mapfile">mapfile</h4>
+
+The mapfile builtin reads lines from stdin into an array
```
Hm this only has `-t` implemented

https://github.com/oilshell/oil/blob/master/frontend/flag_def.py#L118

I would omit the (unimplemented) ones altogether (or somehow comment them out).  Because they might not be implemented for a long time!

dcghub replied:
I have added a usage line and made sure that the options existed. I also have added a few more commands!

andychu commented:
I guess to avoid getting caught up in open-ended designs for the format, I would merge something that only describes documented stuff. 

You can generally check just by doing say `mapfile -d` in OSH.

dcghub commented:
I have been adding some commands and I have added addressed some of your suggestions, let me know what's your opinion.

Minor setback: I haven't been able to test if "help foo" works...because the weird UnicodeEncodeError is happening again...

andychu commented:
Hm can you copy and paste the terminal session here?

https://github.com/oilshell/oil/issues/809


And show "git status" and "git diff".  And `which python`, `python -V` , etc.

andychu commented on doc/help.md:
```diff
@@ -580,49 +580,330 @@ PS1=$x.
+The read builtin reads lines from stdin and stores them into a variable. Input
+is split according to the $IFS delimiter.
+
+read FLAGS VAR+
```
I think this should be `FLAGS* VARS*`

Because both flags and variables are optional.  I started using the regex of syntax of:

```
a*  0 or more
a+  1 or more
a?  0 or 1 (optional)
```

dcghub replied:
Oooh, I noticed you were using them and tried to imitate it, but I didn't realize they were regex. That's a nice idea, will change it.

andychu commented on doc/help.md:
```diff
@@ -580,49 +580,330 @@ PS1=$x.
+terminated by a newline.
 
-<h4 id="read">read</h4>
+echo FLAGS STR
```
ditto, should be `FLAGS* STR*`

andychu commented on doc/help.md:
```diff
@@ -580,49 +580,330 @@ PS1=$x.
+running shell and can modify local variables (as opposed to running in a
+subshell, where it can't)
+
+source SCRIPT ARGS+
```
ARGS*

andychu commented on doc/help.md:
```diff
@@ -580,49 +580,330 @@ PS1=$x.
+
+The cd builtin changes the working directory.
+
+cd FLAGS DIR
```
FLAGS* DIR?

DIR can also be `-` I believe

andychu commented on doc/help.md:
```diff
@@ -580,49 +580,330 @@ PS1=$x.
 
-Reads lines from stdin into a variable -- $REPLY by default.
+ARRAY:
+  Array where the contents will be stored.
```
The description of the array feels a bit redundant.  While there should be consistency, it shouldn't be like "filling out a form".

Also there is a default value `MAPFILE`

dcghub commented:
OK, here is a new iteration, I think I have addressed all your points, let me know if I did some mistake

As for the build problem...I swear I'm not joking, it now works again. I haven't rebooted and/or updated or changed anything AFAIK. I have tried to locate the problem in different branches or past commits, found nothing.

andychu commented:
Great thanks for doing this.   This helps get things going.... I am going to test it out and make another pass over it myself.

andychu commented:
Also I think the unicode issue must be that your editor is inserting Unicode characters like spaces or quotes ?  What editor are you using?

Again I would try `file` right when  you get the error.  It should be possible to at least figure out which file it's coming from

andychu commented:
I made an editing pass over read and echo

https://github.com/oilshell/oil/commit/1994d280b2138e5eeaeeb62989b900df7acda66a

It now renders better as markdown:

http://www.oilshell.org/preview/doc/help.html#read

The dual markdown/text thing is a little weird... I'm not sure if it will hold up, but so far it seems to.  That is, we can render the same text as HTML in the browser and in the terminal.  The conventions are sort of evolving, but it seems that the 4 space block is good for aligned flag descriptions.  And for the usage lnie.

andychu commented:
Some more edits:

https://github.com/oilshell/oil/commit/1de4b9f7776f836ffecf48d81c6bd482c15a7db9

I like how this is turning out!  More PRs in the same direction are appreciated :-)

dcghub commented:
Nice, thanks for the merge! I will take a look at your edits to follow your suggestions in the future.

dcghub commented:
I have some clues about the building problem - I will post them in the bug issue
